new mono drivingfunction handling

fs446 opened the pull request:
I'd make a suggestion to improve readability and code structure w.r.t. our driving function handling. 

Instead of current
`d, selection, secondary_source = sfs.mono.drivingfunction.wfs_25d_point(omega, array.x, array.n, xs)`
it changes to
`d, selection, secondary_source = sfs.mono.wfs.point_25d(omega, array.x, array.n, xs)`
and such like.

The introduced change splits the drivingfunction module into the specific SFS approaches, at the moment `wfs`, `sdm`, `esaedge`, `nfchoa`.
Meaningful for further progress could be the new modules `beamforming` and `panning` covering all aspects of farfield beams and pure gain / pure delay based SFS.

I'm still figuring what the best handling for `def secondary_source_point(omega, c):` would be. At the moment they appear in every SFS-module. This should be solved consistently with the time domain as well.

spors commented:
I would vote for the suggested changes!

mgeier commented on sfs/mono/__init__.py:
```diff
@@ -3,14 +3,23 @@
+    esaedge
+    nfchoa
+    sdm
+    wfs
```
This is alphabetic?

Shouldn't it be somehow "by importance"?

Or probably chronologically with respect to the SFS-Toolbox development?

hagenw replied:
Alphabetic is the easiest to understand and you don't have to fight about which is more important.

fs446 replied:
alphabetic was indeed intentional. I don't have a preference.

mgeier commented on sfs/mono/__init__.py:
```diff
@@ -3,14 +3,23 @@
+    sdm
+    wfs
+
     source
```
What about placing `source` before the driving functions?

fs446 replied:
yes, that seems to be a good categorizing style.

mgeier commented on sfs/mono/__init__.py:
```diff
@@ -3,14 +3,23 @@
+    wfs
+
     source
     soundfigure
```
I'm not sure if it should happen within this PR, but what about moving this to the `wfs` module?

IOW, move `sfs.mono.soundfigure.wfs_3d_pw()` to `sfs.mono.wfs.soundfigure_something()`?

fs446 replied:
moved to sfs.mono.wfs.soundfigure_2d()
I am not sure, it might even be a _3d scenario, can you @spors validate naming.

Note: this code is not tested at all in the moment, so I'm going to open a new issue for this. But for now the handling would be more consistent.

mgeier commented on sfs/mono/esaedge.py:
```diff
@@ -0,0 +1,400 @@
+
+.. include:: math-definitions.rst
+
+.. plot::
```
None of this code is used in this file, it should be completely removed.

fs446 replied:
da37636be625e9dfbb5fa98ceee3b2e336037006

mgeier commented on sfs/mono/nfchoa.py:
```diff
@@ -0,0 +1,259 @@
+    plt.rcParams['figure.figsize'] = 6, 6
+
+    xs = -1.5, 1.5, 0
+    xs_focused = -0.5, 0.5, 0
```
`xs_focused` is unused.

fs446 replied:
23c4f8e575f3d7a63283412ce7de1e43d55d347d

mgeier commented on sfs/mono/nfchoa.py:
```diff
@@ -0,0 +1,259 @@
+    # normal vector for plane wave:
+    npw = sfs.util.direction_vector(np.radians(-45))
+    # normal vector for focused source:
+    ns_focused = sfs.util.direction_vector(np.radians(-45))
```
`ns_focused` is unused.

fs446 replied:
23c4f8e575f3d7a63283412ce7de1e43d55d347d

mgeier commented on sfs/mono/esaedge.py:
```diff
@@ -0,0 +1,400 @@
+"""
+
+import numpy as np
+from numpy.core.umath_tests import inner1d  # element-wise inner product
```
Unused `import`.

fs446 replied:
3779e598b626bb148d4c267c7bc34973583ce772

mgeier commented on sfs/mono/esaedge.py:
```diff
@@ -0,0 +1,400 @@
+from numpy.core.umath_tests import inner1d  # element-wise inner product
+from scipy.special import jn, hankel2
+from .. import util
+from .. import default
```
Unused `import`.

fs446 replied:
3779e598b626bb148d4c267c7bc34973583ce772

mgeier commented on sfs/mono/nfchoa.py:
```diff
@@ -0,0 +1,259 @@
+"""
+
+import numpy as np
+from numpy.core.umath_tests import inner1d  # element-wise inner product
```
Unused `import`.

fs446 replied:
0615628109c14ffd53dce6506b1daa9af6e59e71

mgeier commented on sfs/mono/nfchoa.py:
```diff
@@ -0,0 +1,259 @@
+
+import numpy as np
+from numpy.core.umath_tests import inner1d  # element-wise inner product
+from scipy.special import jn, hankel2
```
`jn` is unused.

fs446 replied:
0615628109c14ffd53dce6506b1daa9af6e59e71

mgeier commented on sfs/mono/nfchoa.py:
```diff
@@ -0,0 +1,259 @@
+from numpy.core.umath_tests import inner1d  # element-wise inner product
+from scipy.special import jn, hankel2
+from .. import util
+from .. import default
```
Unused `import`.

fs446 replied:
0615628109c14ffd53dce6506b1daa9af6e59e71

mgeier commented on sfs/mono/sdm.py:
```diff
@@ -0,0 +1,284 @@
+    plt.rcParams['figure.figsize'] = 6, 6
+
+    xs = -1.5, 1.5, 0
+    xs_focused = -0.5, 0.5, 0
```
Unused.

fs446 replied:
0c86e4d1bfc1a50c12e8f1f0ab18b7715121ac70

mgeier commented on sfs/mono/sdm.py:
```diff
@@ -0,0 +1,284 @@
+    # normal vector for plane wave:
+    npw = sfs.util.direction_vector(np.radians(-45))
+    # normal vector for focused source:
+    ns_focused = sfs.util.direction_vector(np.radians(-45))
```
Unused.

fs446 replied:
0c86e4d1bfc1a50c12e8f1f0ab18b7715121ac70

mgeier commented on sfs/mono/sdm.py:
```diff
@@ -0,0 +1,284 @@
+"""
+
+import numpy as np
+from numpy.core.umath_tests import inner1d  # element-wise inner product
```
Unused.

mgeier commented on sfs/mono/sdm.py:
```diff
@@ -0,0 +1,284 @@
+
+import numpy as np
+from numpy.core.umath_tests import inner1d  # element-wise inner product
+from scipy.special import jn, hankel2
```
`jn` is unused.

mgeier commented on sfs/mono/sdm.py:
```diff
@@ -0,0 +1,284 @@
+from numpy.core.umath_tests import inner1d  # element-wise inner product
+from scipy.special import jn, hankel2
+from .. import util
+from .. import default
```
Unused.

mgeier commented on sfs/mono/wfs.py:
```diff
@@ -0,0 +1,582 @@
+
+import numpy as np
+from numpy.core.umath_tests import inner1d  # element-wise inner product
+from scipy.special import jn, hankel2
```
`jn` is unused.

mgeier commented on sfs/mono/wfs.py:
```diff
@@ -0,0 +1,582 @@
+from numpy.core.umath_tests import inner1d  # element-wise inner product
+from scipy.special import jn, hankel2
+from .. import util
+from .. import default
```
Unused.

mgeier commented:
Looks like you didn't delete `sfs/mono/drivingfunction.py`!

hagenw commented:
> Thanks, this looks really nice!
> 
> I would put the secondary source functions directly into the `sfs.mono` namespace, i.e. define them in `sfs/mono/__init__.py`, right next to `sfs.mono.synthesize()`.

I wasn't aware of this before. I thought `__init__.py` is only for imports or in some cases even completely empty. I wouldn't have looked for any function definitions there.

Do you put all stuff there that doesn't fit into another file?

fs446 commented on sfs/mono/sdm.py:
```diff
@@ -0,0 +1,284 @@
+
+        array = sfs.array.linear(32, 0.2, orientation=[0, -1, 0])
+        d, selection, secondary_source = (
+            sfs.mono.sdm._25d(
```
weird, thought I've completely checked the sphinx output, very obviously not, thx.

fs446 commented on sfs/mono/sdm.py:
```diff
@@ -0,0 +1,284 @@
+        array = sfs.array.linear(32, 0.2, orientation=[0, -1, 0])
+        d, selection, secondary_source = (
+            sfs.mono.sdm._25d(
+                omega, array.x, array.n, npw, [0, -1, 0]))
```
fixed to suggestion

mgeier commented:
@hagenw 

> I wasn't aware of this before. I thought `__init__.py` is only for imports or in some cases even completely empty. I wouldn't have looked for any function definitions there.

Well all non-module (and non-package) contents of a Python "package" (i.e. defined by a directory) has to come from `__init__.py`. Whether it's imported from somewhere else or defined right there doesn't really make a difference.

> Do you put all stuff there that doesn't fit into another file?

I think that's a matter of taste. Some people define their stuff right there, but other make separate submodules (often "hidden", like `_stuff.py`) and then import everything with

```python
from ._stuff import *
```

Neither of those methods are really nice, but that's how it is in Python ...

I slightly prefer putting definitions directly into `__init__.py`, if only to avoid having to come up with a name for a "hidden" submodule.

mgeier commented:
Just one random example for an `__init__.py` from CPython itself: https://github.com/python/cpython/blob/master/Lib/collections/__init__.py

hagenw commented:
OK, wasn't completely aware of the `__init__` stuff.
And you are right, none of the solutions are super nice.

fs446 commented:
With the new handling we could adapt the init source code within the docstring examples to the specific SFS approach, such as SDM gets a linear SSD and so on.

mgeier commented:
@fs446 Yeah that's a good idea, but probably for a separate, later PR?

fs446 commented:
> OK, wasn't completely aware of the `__init__` stuff.
> And you are right, none of the solutions are super nice.

I'm not a big friend of it either, but I'm sure we find an elegant place for it when the changes evolve. At the moment the handling for most user friendly API is good, I like the API very much now.

We furthermore should think about the SFS categories as well:
As I discussed today with @narahahn basically all Mode Matching approaches, such as SDM/NFC-HOA are equivalent with the Equivalent Scattering Approach for the specific SSD type. So in fact we currently dealing with three ESA types with weird names ;-) (SDM, NFC-HOA, Edge) and the WFS type as ray-based/HF/Far Approx of the explicit solutions rather independent of a SSD geometry.
For the moment I'd vote to go with the currently proposed name scheme, but as soon as we implement other panning, beamforming stuff we should seriously think about this categorizing.

narahahn commented:
First of all, thank's a lot for the changes. I really like them.

> As I discussed today with @narahahn basically all Mode Matching approaches, such as SDM/NFC-HOA are equivalent with the Equivalent Scattering Approach for the specific SSD type. So in fact we currently dealing with three ESA types with weird names ;-) (SDM, NFC-HOA, Edge) and the WFS type as ray-based/HF/Far Approx of the explicit solutions rather independent of a SSD geometry.
> For the moment I'd vote to go with the currently proposed name scheme, but as soon as we implement other panning, beamforming stuff we should seriously think about this categorizing.

For now, I'd also keep the current naming scheme.
Despite the similarity among the mode-matching approaches,
the implementations have been evolved and optimized in different ways
mostly due to the specific SSDs under consideration.
IMHO, each deserves its own submodule.

I have one suggestion, though.
How about change `esaedge` to `esa` and
put the driving function for edge SSD under the module?  Like `esa.point_edge()`.
Someday, we might want to add other `esa` driving functions for different types of SSD,
that are not covered by `sdm` and `nfchoa`.

mgeier commented on sfs/mono/sdm.py:
```diff
@@ -0,0 +1,263 @@
+        :context: close-figs
+
+        array = sfs.array.linear(32, 0.2, orientation=[0, -1, 0])
+        d, selection, secondary_source = (
```
```python
        d, selection, secondary_source = sfs.mono.sdm.point_25d(
            omega, array.x, array.n, xs, [0, -1, 0])
```

fs446 replied:
aed7ab787c18b904e3d39f08a1b5f8725ee68596

mgeier commented:
In `sfs/mono/__init__.py` I would move each of the `import` statements either to the top or the bottom (as mentioned in individual comments).

Then you should be free to choose any ordering for the functions.

I would not define `secondary_source_point()` or `secondary_source_line()` first, because they are the most boring functions in this module.

fs446 commented:
Idea from @narahahn  
> I have one suggestion, though.
> How about change `esaedge` to `esa` and
> put the driving function for edge SSD under the module? Like `esa.point_edge()`.
> Someday, we might want to add other `esa` driving functions for different types of SSD,
> that are not covered by `sdm` and `nfchoa`.

is great. I've changed esaedge to esa in 7424338f93fbd01cdf1aa6f23f97895b06c6cada
renamed the functions, which is even more consistent now and left some
docstring comments for better understanding what is/should going on in this module.
